doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+
+These rates are _source frontier differences_. Each source's `SourceReader` implementation will
+be required to expose a continuous stream of _upstream source frontiers_. This means:
+
+- The Kafka source will periodically fetch metadata and expose per-partition offset high-watermarks.
+- The MySQL source will periodically expose the result of `SELECT @gtid_executed`, which is the gtid of the latest transaction.
+- The Postgres source will periodically expose the result of `SELECT pg_current_wal_lsn()`, which is the lsn of the latest transaction.
+
+Additionally, the source pipeline will periodically invert the latest frontier we have committed for the source, from the Materialize timestamp
+domain to the source-timestamp domain. For MySQL and Postgres sources, this frontier will the `meet` of the subsource frontiers.
+
+`upstream_rate` will be derived from the two latest _upstream source frontiers_, by calculating _source frontier difference_ between them.
+Similarly `process_rate` will be derived from the two latest committed frontiers.
+
+
+### Existing metrics
+
+Existing metrics such as `messages_received` and `bytes_received` continue to be useful for understanding the general performance of a given source.
+If users find some of them confusing or not useful, we can remove them in the future.
+
+## Minimal Viable Prototype
+
+N/A for now. This design document is primarily designed to expose
+
+
+## Alternatives
+
+- Expose 100% accurate information about how much of the snapshot we have processed, as opposed to how much we have read.
+    - Rejected as it's considered as difficult as implementing Hydration Backpressure.
+- Expose progress information in a different format per source type.
+    - Rejected as it adds undue complexity in the system catalog, console, and for the user to interpret.
+        - In the future, if customers want per-partition information for Kafka sources, we can do so easily using this same general design.
+- Consolidate the two metric sets proposed in this design document into one.
+    - Rejected due to its complexity.
+- Expose more information to users about their sources.
+    - This design document is intentionally constrained, to help users answer 2 well-known questions about their sources.
+
+
+## Open questions
+
+- Are there any edge cases around _source frontier differences_ that can't be reconciled into a relatively meaningful number for the purpose of metrics?
+- Are there other scenarios than the ones predicted in the above charts that we aren't capturing?
+    - One case might be: people's sources that fall behind upstream just use more memory, while Materialize continues to read as fast as possible.
+    The metrics added by this proposal would allow us to easily distinguish between those cases.
+- Should `upstream_rate` and `process_rate` be exposed as rates (as described in this design document), or should we instead expose the maximum value, and
+leaving calculating the rate to the client?
```

Comment:
I like the sound of exposing the maximum value. Seems more in line with the way the other statistics table work? For everything else in the table, the client has to calculate the rate themselves.

Reply:
Sounds good, ill update this document